modules/standard/Heap.chpl

```diff
@@ -0,0 +1,430 @@
+      _heapify_up(_data.size-1);
+    }
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _push(element);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list in which elements are to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      _enter();
+      for e in x do
+        _push(x);
+      _leave();
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array in which elements are to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      _enter();
+      for e in x do
+        _push(e);
+      _leave();
+    }
+
+    /*
+      Pop an element and return it.
+
+      :return: the top element
+      :rtype: eltType
+    */
+    proc pop(): eltType {
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.pop\" on an empty heap.");
+      }
+
+      if _data.size != 1 then
+        _data(0) <=> _data(_data.size-1);
+
+      var ret = _data.pop();
+      _heapify_down(0);
+      _leave();
+      return ret;
+    }
+
+    /*
+      Iterate over the elements of this heap
+
+        .. note::
+          Elements are not yielded in order
+
+    */
+    iter these() ref {
+      for e in _data {
+        yield e;
+      }
+    }
+
+    /*
+      Iterate over the elements of this heap
+      
+        .. note::
+          Elements are yielded in order and removed from the heap
+    */
+    iter consume() {
+      var h = this;
+      while !h.isEmpty() {
+        yield h.pop();
+      }
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this heap.
+
+        .. note::
+          Elements of the array are not in order
+
+      :return: A new DefaultRectangular array.
+    */
+    proc const toArray(): [] eltType {
+      if !isCopyableType(eltType) then
+        compilerError("toArray() method is not avaliable on a 'heap'",
+                      " with elements of a type that can't be copied, here: ",
+                      eltType: string);
+      _enter();
+      var l = _data.toArray();
+      _leave();
+      return l;
+    }
+
+    /*
+      Write the contents of this heap to a channel.
+
+        .. note::
+          Elements are not written in order
+
+      :arg ch: A channel to write to.
+    */
+    proc writeThis(ch: channel) throws {
+      _enter();
+      ch <~> this._data;
+      _leave();
+    }
+  }
+
+  /*
+    Create a heap from a list.
+
+    :arg x: The list to initialize the heap from.
+    :type x: `list(?t)`
+
+    :arg comparator: The comparator to use
+
+    :rtype: heap(t, comparator)
+  */
+  proc createHeap(x: list(?t), parSafe: bool = false, comparator = defaultComparator) {
```

Comment:
Can we make the argument `x` a `const ref`?

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/e3aba540cfd067634389eb70d08f2a7d8038a581